rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,270 @@
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+                ecdh_key.inner,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    let instance = EcDh {
+                        curve_type: curve,
+                        key: ecdh_key,
+                    };
+                    Ok(instance)
+                }
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// EcDh::from_public_key_bytes() returns an EcDh struct that only has a public key attached.
+    pub fn from_public_key_bytes(
+        curve: CurveType,
+        public_key: &[u8],
+    ) -> Result<Self, SymCryptError> {
+        let num_format = get_num_format(curve);
+
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        let edch_key = EcDhKey::new(curve)?;
+
+        unsafe {
+            // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptEckeySetValue(
+                std::ptr::null(),
+                0,
+                public_key.as_ptr(),
+                public_key.len() as symcrypt_sys::SIZE_T,
+                num_format,
+                ec_point_format,
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+                edch_key.inner,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    let instance = EcDh {
+                        curve_type: curve,
+                        key: edch_key,
+                    };
+                    Ok(instance)
+                }
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    pub fn get_public_key_bytes(&mut self) -> Result<Vec<u8>, SymCryptError> {
+        let num_format = get_num_format(self.curve_type);
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        unsafe {
+            // SAFETY: FFI calls
+            let pub_key_len = symcrypt_sys::SymCryptEckeySizeofPublicKey(
+                self.key.inner,
+                symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY,
+            );
+
+            let mut pub_key_bytes = vec![0u8; pub_key_len as usize];
+
+            match symcrypt_sys::SymCryptEckeyGetValue(
+                self.key.inner,
+                std::ptr::null_mut(),
+                0 as symcrypt_sys::SIZE_T,
+                pub_key_bytes.as_mut_ptr(),
+                pub_key_len as symcrypt_sys::SIZE_T,
+                num_format,
+                ec_point_format,
+                0, // No flags allowed
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(pub_key_bytes),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    pub fn ecdh_secret_agreement(private: &EcDh, public: &EcDh) -> Result<Vec<u8>, SymCryptError> {
+        if private.curve_type != public.curve_type {
+            return Err(SymCryptError::InvalidArgument);
```

Comment:
>return Err(SymCryptError [](http://example.com/codeflow?start=12&length=24)

SymCrypt will validate that the curves are the same for you. I think it's probably best to just rely on SymCrypt for that and not duplicate parts of the logic here.